### ISSUE TITLE
Tabs produce a FOIT (Flash of Invisible Text)

### ISSUE THREAD

**Issue by summercms** [octobercms/october#5393]
A few examples to explain the issue.

![image](https://user-images.githubusercontent.com/57409060/101645943-d7e5cb80-3a2e-11eb-8335-78b56306fedc.png)

Above: When opening tab one `directly` there is no error.

![image](https://user-images.githubusercontent.com/57409060/101646051-f64bc700-3a2e-11eb-96bd-319e7d851613.png)

Above: When opening tab two `directly`, October displays the whole contents of tab one before opening tab two! If there is a lot of data in tab one this can be shown for quite a long time. Especially when a developer has added a custom widget to tab one.

### Steps to reproduce this error

Open up a tab directly (excluding tab 1), e.g.

```
http://localhost/backend/acme/example/something/update/1#primarytab-tab2
```

### Actual behaviour

When I load a page in my browser I see the form field widgets being loaded above the tabs, then the tabs getting loaded, I see tab one being loaded first (being displayed for several hundred milliseconds) then tab two being loaded.

The more form fields being added to tab one the worse the problem gets.

### Expected behaviour

When I load a page in my browser I see the form field widgets being loaded above the tabs, then the tabs getting loaded and I only see tab two (no flash of form fields from tab one).

**Comment by mjauvin** [comment]
The anchor is applied client-side, so I don't really see a way around this, unless you apply custom javascript.

**Comment by ayumi-cloud** [comment]
In my browser I see the form field widgets being loaded above the tabs, then the tabs getting loaded, seeing tab one being loaded first (maybe several hundred milliseconds) then tab two being loaded.

Using a modern tab components from other libraries you don't see this error. I was hoping we could add the same technology to October to clean up the tabs.

**Comment by ayumi-cloud** [comment]
In my browser I see the form field widgets being loaded above the tabs, then the tabs getting loaded, seeing tab one being loaded first (maybe several hundred milliseconds) then tab two being loaded.

Using a modern tab component from other libraries you don't see this error. I was hoping we could add the same technology to October to clean up the tabs.

p.s. Was surprised to see this error as the tabs are meant to be lazy loaded.

**Comment by ayumi-cloud** [comment]
Going to play around with this issues.

I'm thinking of setting tab one's css to `display: none` on page load and then the javascript can display the tab requested tab and hopefully this gets rid of this annoying flash of undesired code and speed up loading?

**Comment by ayumi-cloud** [comment]
Going to play around with this issue.

I'm thinking of setting the css for the class `tab-content` to `display: none` on page load and then the javascript can display the requested tab's content and hopefully this gets rid of this annoying flash of undesired code and speed up loading by not having to style two sets of tabs.

https://github.com/octobercms/october/blob/0718d04a7f3df0e3e40771ae96b8627a18cb7f1d/modules/backend/widgets/form/partials/_form_tabs.htm#L40

**Comment by ayumi-cloud** [comment]
I don't think the answer needs javascript.

File location: https://github.com/octobercms/october/blob/0718d04a7f3df0e3e40771ae96b8627a18cb7f1d/modules/backend/widgets/form/partials/_form_tabs.htm#L40-L56

First sets the variable `index` to position one:

```php
$index = 0;
```

Then just echo's out the class at position one:

```php
<?= ($index++ === 0) ? 'active' : '' ?>
```

There's no clever logic to this!

Instead, the cms could check the `url` to see if it contains `#primarytab-` etc. and if it does then don't display the class `active` at position one.

.. then I take it, the javascript later on inserts the `active` class into the correct element location.

**Comment by ayumi-cloud** [comment]
I see what you mean with the `fragment` url, needing javascript.

File location: https://github.com/octobercms/october/blob/0718d04a7f3df0e3e40771ae96b8627a18cb7f1d/modules/backend/widgets/form/partials/_form_tabs.htm#L40-L56

### Current method

First the cms sets the variable `index` to position one:

```php
$index = 0;
```

Then just echo's out the class at position one:

```php
<?= ($index++ === 0) ? 'active' : '' ?>
```

There's no clever logic to this!

### Updated method

1. Remove that php code from the cms.

2. Use javascript `window.location.hash` to get the `fragment` url and check if it contains `primarytab-` or `secondarytab-`

```js
var $fragment = window.location.hash;

if ( (!$fragment.includes("primarytab-")) || (!$fragment.includes("secondarytab-")) ) {
    // Didn't match the fragment so add the `active` to position one
    $('#Form-primaryTabs .tab-content li:first').addClass('active');
    $('#Form-secondaryTabs .tab-content li:first').addClass('active');
}
```

**Comment by summercms** [comment]
I see what you mean with the `fragment` url, needing javascript.

File location: https://github.com/octobercms/october/blob/0718d04a7f3df0e3e40771ae96b8627a18cb7f1d/modules/backend/widgets/form/partials/_form_tabs.htm#L40-L56

### Current method

First the cms sets the variable `index` to position one:

```php
$index = 0;
```

Then just echo's out the class at position one:

```php
<?= ($index++ === 0) ? 'active' : '' ?>
```

There's no clever logic to this!

### Updated method

1. Remove that php code from the cms.

2. Use javascript `window.location.hash` to get the `fragment` url and check if it contains `primarytab-` or `secondarytab-`

```js
var $fragment = window.location.hash;

if ((document.querySelector('#Form-primaryTabs') !== null) && (!$fragment.includes("primarytab-"))) {
    // Didn't match the fragment so add the `active` to position one
    $('#Form-primaryTabs .tab-content li:first').addClass('active');
}

if ((document.querySelector('#Form-secondaryTabs') !== null) && (!$fragment.includes("secondarytab-"))) {
    // Didn't match the fragment so add the `active` to position one
    $('#Form-secondaryTabs .tab-content li:first').addClass('active');
}
```

3. I take it October then adds the `active` class to the non-tab one already.

**Comment by mjauvin** [comment]
Care to submit a PR @ayumi-cloud ?

**Comment by bennothommo** [comment]
@ayumi-cloud A potential fix is to remove the PHP code that applies the active state, and simply apply the active state through JavaScript (either the required tab from the hashbang, or the first tab if no hashbang is available). That way, only one tab is made active on initialisation.

**Comment by ayumi-cloud** [comment]
What bennothommo said is what I came to the conclusion as well.

---

I should mention a related pr is found here: https://github.com/octobercms/october/pull/4666

But it should be noted that pr doesn't fix this issue!

---

A couple of things I was thinking in the back of my mind:

1. I was thinking there could be an issue if someone tried to load `tab one` with a **fragment/hashbang** directly, e.g.

```
http://localhost/backend/acme/example/something/update/1#primarytab-tab1
```

So would need to add some extra code for that.

2. October uses Bootstrap, from my research yesterday I noticed that Bootstrap seems to add **two** active classes, one for the `tab` and one for the `tab content`. 

https://github.com/octobercms/october/blob/0718d04a7f3df0e3e40771ae96b8627a18cb7f1d/modules/backend/widgets/form/partials/_form_tabs.htm#L19

https://github.com/octobercms/october/blob/0718d04a7f3df0e3e40771ae96b8627a18cb7f1d/modules/backend/widgets/form/partials/_form_tabs.htm#L47

Let me grab a gif I saw with the first lazyloading pr, link: https://github.com/octobercms/october/pull/4658

![66123990-944e3100-e5e3-11e9-81fc-7a8813809972](https://user-images.githubusercontent.com/57409060/101784980-20b38800-3af4-11eb-9289-cfe7af6b0798.gif)

Breaking down the above gif. Let's say a user is on a using a **2G/GPRS network** connection and the data is coming very slowly to their device:

- First - The `tab` list loads

![image](https://user-images.githubusercontent.com/57409060/101786106-7472a100-3af5-11eb-8a5d-259f9e53ec38.png)

- Second - You see a blue loading spinner for the `tab content`

![image](https://user-images.githubusercontent.com/57409060/101786145-80f6f980-3af5-11eb-87a6-857815b29ce4.png)

- Third - The `tab content` gets loaded.

![image](https://user-images.githubusercontent.com/57409060/101786200-91a76f80-3af5-11eb-9015-29c075e90f1e.png)

This should be the loading sequence for the tabs in the backend.

Please note, when a user is using a **fast internet connection** all these sequences would take **milliseconds**.

Basically I'm saying the javascript code adjusting the `active` class should output this loading sequence, for when the page loads for the **first-time!**

**Comment by ayumi-cloud** [comment]
https://github.com/octobercms/october/blob/develop/modules/backend/widgets/form/assets/js/october.form.js

**Comment by ayumi-cloud** [comment]
https://github.com/octobercms/october/blob/2d6b35b31e05a61b2f042a8f5d4fac85b09ef629/modules/backend/widgets/form/assets/js/october.form.js#L235-L242

**Comment by ayumi-cloud** [comment]
@mjauvin @bennothommo 

Why are we suggesting to check the `fragment` url? Seems pointless! Why not remove the php and just use javascript to add the `active` class to the first `visible tab-content` on page-load instead.

**Comment by ayumi-cloud** [comment]
https://github.com/octobercms/october/blob/develop/modules/backend/widgets/form/assets/js/october.form.js

**Comment by ayumi-cloud** [comment]
@mjauvin 

> Care to submit a PR @ayumi-cloud ?

Just working on it...

**Comment by ayumi-cloud** [comment]
https://github.com/octobercms/october/blob/develop/modules/backend/widgets/form/assets/js/october.form.js
